system_tests/common_test.go

```diff
@@ -60,6 +60,128 @@ import (
 type info = *BlockchainTestInfo
 type client = arbutil.L1Interface
 
+type NodeBuilder struct {
+	// Nodebuilder configuration
+	ctx           context.Context
+	chainConfig   *params.ChainConfig
+	cacheConfig   *core.CacheConfig
+	nodeConfig    *arbnode.Config
+	l1StackConfig *node.Config
+	l2StackConfig *node.Config
+	isSequencer   bool
+
+	// L1 Nodebuilder fields
+	L1Info    info
+	L1Client  *ethclient.Client
+	L1Backend *eth.Ethereum
+	L1Stack   *node.Node
+
+	// L2 Nodebuilder fields
+	L2Info       info
+	L2Client     *ethclient.Client
+	L2Backend    *eth.Ethereum
+	L2Node       *arbnode.Node
+	L2Stack      *node.Node
+	L2ChainDB    ethdb.Database
+	L2NodeDB     ethdb.Database
+	L2Blockchain *core.BlockChain
+}
+
+func NewNodeBuilder(ctx context.Context) *NodeBuilder {
+	return &NodeBuilder{ctx: ctx}
+}
+
+func (b *NodeBuilder) SetChainConfig(c *params.ChainConfig) *NodeBuilder {
+	b.chainConfig = c
+	return b
+}
+
+func (b *NodeBuilder) SetNodeConfig(c *arbnode.Config) *NodeBuilder {
+	b.nodeConfig = c
+	return b
+}
+
+func (b *NodeBuilder) SetCacheConfig(c *core.CacheConfig) *NodeBuilder {
+	b.cacheConfig = c
+	return b
+}
+
+func (b *NodeBuilder) SetL1StackConfig(c *node.Config) *NodeBuilder {
+	b.l1StackConfig = c
+	return b
+}
+
+func (b *NodeBuilder) SetL2StackConfig(c *node.Config) *NodeBuilder {
+	b.l2StackConfig = c
+	return b
+}
+
+func (b *NodeBuilder) SetL1Info(l1Info info) *NodeBuilder {
+	b.L1Info = l1Info
+	return b
+}
+
+func (b *NodeBuilder) SetL2Info(l2Info info) *NodeBuilder {
+	b.L2Info = l2Info
+	return b
+}
+
+func (b *NodeBuilder) SetIsSequencer(v bool) *NodeBuilder {
+	b.isSequencer = v
+	return b
+}
+
+func (b *NodeBuilder) CreateTestNodeOnL1AndL2(t *testing.T) *NodeBuilder {
+	b.L2Info, b.L2Node, b.L2Client, b.L2Stack, b.L1Info, b.L1Backend, b.L1Client, b.L1Stack =
+		createTestNodeOnL1AndL2Impl(t, b.ctx, b.isSequencer, b.nodeConfig, b.chainConfig, b.l2StackConfig, b.L2Info)
+	return b
+}
+
+func (b *NodeBuilder) CreateTestNodeOnL2Only(t *testing.T, takeOwnership bool) *NodeBuilder {
```

Comment:
instead of CreateTestNodeOnX - I would go for spltting config and create phase.
So you could have e.g. ConfigForL2() vs Config for L2OnL1(), an a build() step that just uses configs created earlier - with the advantage that now the test can modify the default config between these two steps.
same sort of goes for isSequencer.. As much as possible should be done on creating configs on one hand, then possibly test overrides them, then build everything according to these configs.